계산기 I [STEP 1] 바드

bar-d opened the pull request:
@GREENOVER   
안녕하세요 그린! 첫 PR 잘부탁드립니다🙇‍♂️  
아무래도 `Queue`, `TDD`, `UML`등 익숙치 않고 처음 보는 개념들이 있어 어렵게 느껴졌습니다. 
많은 조언 부탁드리겠습니다

## `UML`
![Screen Shot 2022-05-18 at 5 34 50 AM](https://user-images.githubusercontent.com/92622931/168905339-91963e24-8dae-4975-bea8-ccc92a10c135.png)

---
## `queue`
`queue`타입을 구현하기 위해서는 적어도 4가지 기능이 있어야 했는데
 - enqueue() 값을 넣을 수 있는 메소드
- dequeue() 값을 꺼낼 수 있는 메소드
- peek 프로퍼티를 통해 다음으로 나올 값을 알수 있어야 함
- isEmpty 프로퍼티를 통해 큐가 비어있는지 알 수 있어야 함
가 있어야 한다고 보았고 이 것을 토대로 `Queue`라는 프로토콜을 만들어주었습니다.

## `queue 구현 방법`
처음으로 `queue`라는 것을 공부하게 되어서 `queue`에대해 알아보니, 제가 찾아본 대표적인 방법들은 네가지가 있었습니다.
### 1. `Array`를 이용해 `queue`를 구현하는 방법
- 빈 공간에 추가할 땐 수행속도가 빠름 
- 가끔식 크기가 늘려야 할 때 수행속도가 O(n) -> 배열의 메모리를 n번 반복해서 할당
- 스위프트 배열 구조체의 특징
- 배열에서 deQueue를 하게 되면 한 칸씩 앞으로 이동시켜야 하기 때문에 수행시간이 O(n)이 됨 -> 비효율적인 면이 있음
- 이러한 단점 때문에 성능이 낮아지므로 큐를 다르게도 구현
### 2. `DoubliylLinkedList Queue`
- 연결리스트의 확장된 개념
- 연결리스트는 자기 자신의 노드의 값과 다음 노드를 가리키는 `참조 정보`로 구성되지만 여기서 더 나아가서 현재 노드가 다음 노드의 `참조 정보`뿐만 아니라 이전 노드의 `참조 정보`도 가지고 있는 것이 더블 연결리스트   

	- `장점`

		- 이제 deQueue 메소드의 수행속도가 O(1)로 빠름 (ArrayQueue는 Q(n))
		- 이는 연결리스트의 특징인 아이템이 제거되도 나머지 요소들이 재정렬하지 않아도 되기 때문  
	- `단점`
		- 연결리스트를 추가할 때 수행하는 동적할당이 상대적으로 큰 메모리를 소모
		- 값, 이전노드 참조, 다음노드 참조 이렇게 3개를 포함

### 3. `DoubleStack Queue`
- 배열을 사용해 `Queue`를 구현하면 큐에서 하나를 뺄때마다 배열을 앞으로 옮겨줘야함 -> 재정렬이 필요
- 더블 스택은 처음 저장한 스택을 반대로 저장해놓은 스택을 가지고 있기에 반대로 저장해놓은 스택에서 마지막 요소를 제거해주면 재정렬이 필요하지 않으므로 시간복잡도가 O(1)
- LeftStack: deQueue 전용 스택 deQueue를 수행할 아이템들은 LeftStack에 들어감
- RightStack: enQueue 전용 스택 enQueue를 수행할 아이템들은 Right Stack에 들어감
- enQueue를 수행할 RightStack에 스택 자료구조의 append 메소드를 사용해서 값을 추가
- deQueue를 수행하면 enQueue전용인 RightStack의 요소들을 거꾸로 정렬한 뒤 LeftStack에 옮기고 popLast() 메소드를 사용해 마지막 값을 제거
- 거꾸로 정렬하는 이유는 큐의 성질인 먼저 들어온 아이템을 먼저 제거하면서 수행속도까지 빠르게 하기 위함
- 거꾸로 정렬한 왼쪽 스택에서 마지막 요소를 제거하면 가장 먼저들어온 것이 제거됨
- 배열로 구현한 큐와 비교하면
	- deQueue() 메소드의 수행속도가 빠름(O(1) > O(n))
- 더블 연결리스트로 구현한 큐(QueueLinkedList)와 비교하면 
	- 공간복잡성 측면에서 연결리스트 보다 더 좋음
	- 더블 스택으로 구현한 큐의 공간 복잡성: 빠름
	- 연결 리스트로 구현한 큐의 공간 복잡성: 메모리 접근 시간이 느림
	- 연결리스트의 특징인 참조 정보로 구성되기 때문에 메모리에 접근하는 시간이 더블스택에 비해 오래 걸림

### 4. `RingBuffer Queue`
- 고정된 크기의 배열을 사용하고, 배열이 큐의 값들로 가득 차면 다시 첫 번째 아이템으로 돌아오는 방법을 사용
- 더블 연결리스트와 더블 스택으로 구현한 큐와 실행속도는 변다른 차이가 없지만 고정된 배열을 사용해야 한다는 단점이 있음

### ✅`Double Stack`
- *이 네가지 방법 중에 이해도 제일 잘 됬고 수행 속도면에서도 좋아보이는 더블스택을 사용하여 `Queue`를 구현하였습니다.*


## `시간 복잡도`
- 해당 STEP에서 사용한 Array 시간복잡도
	- append(_ newElememt: Element): 평균 시간 복잡도는 O(1), 최악의 상황은 메모리를 재할당 할 때 시간복잡도가 O(N)   
	- `isEmpty()`: O(1)   
	- `popLast()`: O(1)   
	- `reversed`: O(1)     
	- `removeAll`: O(N) 
	- `first`: O(N)   
	- `last`: O(1)   
	- `append(_newElement: Element)`: 평균 시간 복잡도는 O(1), 최악의 상황은 메모리를 재할당 할 때 시간복잡도가 O(N)  

## 🙋궁금한 점
- 사실 아직 이 프로젝트의 수행사항을 제대로 이해를 하지 못하고 있는 느낌이 듭니다.
- 아직 `UML`과 `TDD`가 익숙치 않아 제가 제대로 한 것인지를 아직 잘 모르겠습니다.🥲
### `Queue`
- 스위프트에는 `Queue`라는 타입이 없어서 여러 방식으로 구현을 하여 사용을 하는데, `DispatchQueue`같은 클래스는 왜 존재하는지, 특정 케이스에서만 `Queue`가 쓰이도록 구현을 해논건지 궁금합니다.  
아니면 구현을 어떻게 하느냐에 따라 쓰이는 용도가 달라서일까요?



- 해당 프로젝트 사항 중에 `사칙연산 계산을 위한 큐(Queue) 타입(CalculatorItemQueue)과 큐타입 구현을 위한 List 타입을 직접 UML로 표현해봅니다`라는 말이 있는데 여기서 `list`타입이라는 것은 `Array`같이 순서를 가진 타입을 말하는 걸까요?
### `TDD`
- `TDD`를 수행하기 전 먼저 `UML`을 작성해야 했는데, `UML`을 토대로 `TDD`를 작성하려고 하니 아래와 같은 상황에서 시작을 하였습니다.
```swift
struct CalculatorItemQueue<Double>: Queue, CalculateItem {  
    var leftStack: Array<Double> = []
    var rightStack: Array<Double> = []
    
    var isEmpty: Bool
    
    var peek: Double?
    
    
    mutating func enQueue(_ double: Double) -> Bool {
        return true
    }
    
    mutating func deQueue() -> Double? {
        return Double?.self as? Double
    }
}
```
- `TDD`는 말 그대로 테스트를 기반으로 하는 코드인데, `UML`을 이미 작성을 해서 `UML`을 토대로 코드의 틀만 짠 상태로 테스트 코드를 진행하는게 맞는 방법일까요?

bar-d committed:
feat: 파일 정리 및 폴더 생성

bar-d committed:
feat: Protocol 타입인 Queue 생성 및 Foundation 제거

bar-d committed:
feat: CalculatorItemQueue구조체 생성 및 Queue 프로토콜을 구조체에 사용하기 위해 함수 내 mutating 추가

bar-d committed:
feat: TDD를 통해 CalculatorItemQueue 구현

bar-d committed:
feat: 빈 프로토콜인 CalculateItem 생성 및 CalculatorItemQueue에 CalculateItem 채택

bar-d committed:
feat: 중복된 파일 제거 및 폴더 재정렬

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        sut.leftStack = [1.0, 2.0, 3.0]
+        sut.rightStack = [3.0, 2.0]
+        //when
+        func isReturnWhichStack() -> Double {
+            if !sut.leftStack.isEmpty {
+                return sut.leftStack.first ?? -10000.0
+            } else {
+                return sut.rightStack.last ?? -10000.0
+            }
+        }
+        let result = isReturnWhichStack()
```
강제적으로 left/right를 맞춰준거라 유의미한 테스트가 되긴 힘들것 같아요..!
하시고 싶으셨던게 유추하자면 leftStack이 reversed되었을때 left/right의 요소를 비교하는것 같았는데 맞을까요?
현재 구현에서는 무조건 강제적으로 성공시켜버린것 같아서 올바른 큐의 메서드나 속성을 테스트하기엔 무리가 있어보입니다ㅠ
leftStack이 비어있지 않았을때 reverse해서 rightStack하고 비교하는 그림은 어떨까요?🙋🏻

bar-d replied:
`peek`이라는 프로퍼티가 큐의 첫번째 값을 반환해주는 프로퍼티인데, `enQueue`스택에 값이 있다면 `enQueue`스택의 첫번째 값을,   `enQueue`스택에 값이 없다면, `deQueue`의 마지막 값을 반환을 해줘야 하는 프로퍼티 입니다.
해당 테스트는 현재 peek의 프로퍼티가 `enQueue`스택의 값이 있다면 `enQueue`스택의 첫번째 값을 반환하는지 확인을 하는 테스트인데, 제가 테스트 함수명을 잘못 적어놓은 것 같습니다.🥲
이에 `enQueue`스택의 첫번째 값이 `deQueue`의 마지막 값과 비교하는 것이 아니라 `enQueue`스택이 비어있지 않을 때, `enQueue`스택의 첫번째 값을 반환하는지 확인하는 테스트를 만들려고 하였습니다.
혼동을 드려 죄송합니다🙇‍♂️

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        sut.rightStack = [3.0, 2.0]
+        let input = 4.0
+        //when
+        func test_enQueue() {
+            sut.leftStack.append(input)
+        }
+        test_enQueue()
```
이 메서드를 내부에서 구현하고 호출해주신 의도가 있을까요~?🙋🏻
만약 여러번 사용되는 메서드일 경우에는 밖으로 빼는게 나을것 같습니다!
그전에 사실 sut.leftStack.append(input)을 직접 해줌으로써 메서드 구현이 필요없어 보이는데 어떠신가요?🤔

bar-d replied:
처음 `CalculatorItemQueue`를 `UML`을 통해 구현을 했을 때 `enQueue`의 함수 내는 구현이 되어있지 않기에 함수를 뼈대만 가지고 구성을 한다고 생각하여 `test_enQueue()`라는 가상의 메소드를 통하여 똑같이 구현해보고자 함수 내에 메서드를 사용해주었습니다.    
가상 함수 내에 구현을 하지 않고 본 함수 내에서 쓰일 메소드만 적어도 괜찮은건가요?

GREENOVER replied:
네네! TDD를 하면서 현재 테스트의 목적은 UML 기반으로 만들어진 CalculatorItemQueue의 메서드들에 값을 넣어줄때 정상적인지 제 기능을 하는지 판단하기 위함이라 그 방향이 더 적합할것 같습니다!

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+            sut.leftStack.append(input)
+        }
+        test_enQueue()
+        let result = sut.leftStack.last
+        //then
+        XCTAssertEqual(result, 4.0)
```
result 상수로 하나 두고 XCTAssertEqual에서 해당 상수를 사용하는것과
XCAssertEqual(sut.leftStack.last, 4.0)하는것은 어떤 차이가 있을까요?🙋🏻

bar-d replied:
별 다른점은 없어보입니다만, `sut.leftStack.last`의 값의 결과가 4.0과 같아야 하는 것이므로  `result` 상수에는 4.0이라는 숫자를 넣어놓는 것이 더 명확한 테스트라고 볼 수 있을 것 같습니다.  
그린의 생각은 어떠하신가요?

GREENOVER replied:
상수를 하나 만들어주는 메모리까지도 아끼고 싶은 분들도 있을 수 있고 상수를 두어 더 명확하게 가져가고 싶은 분들도 있어서
정말 코드 스타일의 차이라 여쭤봤습니다!
명확한 이유가 있다면 그대로 고하시면 됩니다👍🏻

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        sut.leftStack = [1.0, 2.0, 3.0]
+        sut.rightStack = []
+        //when
+        func test_deQueue() {
+            if sut.rightStack.isEmpty {
+                sut.rightStack = sut.leftStack.reversed()
+            }
+        }
```
이것도 마찬가지로 메서드를 만들 이유가 있을까란 생각이 들었습니다..!

bar-d replied:
위에서 답한 것과 같은 이유로 만들어주었습니다.

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        }
+        
+        test_deQueue()
+        if sut.rightStack.isEmpty {
+            sut.rightStack = sut.leftStack.reversed()
+        }
```
위 구현과 중복인것 같습니다!⚠️

bar-d replied:
테스트 하는 도중 지워주질 못했네요🥲

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        sut.leftStack = [1.0, 2.0, 3.0]
+        sut.rightStack = [3.0, 2.0, 1.0]
+        //when
+        func test_deQueue() {
+            if sut.leftStack.reversed() == sut.rightStack {
+                sut.leftStack.removeAll()
+            }
+        }
```
이것도 마찬가지로 메서드로 만들지 않아도 될것 같아요🤔

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        //given
+        sut.rightStack = [3.0, 2.0, 1.0]
+        //when
+        func test_deQueue() -> Double {
+            return sut.rightStack.last ?? -10000.0
+        }
```
요기도 메서드 구현까지는 필요 없어보입니다🤔

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        //given
+        sut.leftStack = [1.0, 2.0, 3.0]
+        sut.rightStack = []
+        sut.enQueue(4.0)
+        sut.deQueue()
```
실제 enQueue, deQueue는 값을 반환해주는 시그니쳐인데 반환된 값을 어디 담아주거나 사용하지 않아 보이기에 적절한 구현인지 궁금합니다🙋🏻

bar-d replied:
`enQueue` 함수에서 반환하는 값을 없애주었고, `deQueue`는 값을 반환해줘야 하기에 `result`에 넣어주었습니다

GREENOVER commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        sut.leftStack = [1.0, 2.0, 3.0]
+        sut.rightStack = []
+        //when
+        sut.deQueue()
```
요 deQueue 메서드도 반환 타입을 던져주는데 호출부에서 값을 받지 않고 있네요🤔

bar-d replied:
반환타입을 왜 안받아주었을까요🥲
또한 `test_오른쪽스택이_비어있을때_deQueue를호출하면_enQueue스택이_비어지는가()`라는 테스트와 `test_deQueue함수를_호출했을때_왼쪽스택의순서를뒤집어_오른쪽스택에집어넣었으면_왼쪽스택의요소를_모두없애는가()`라는 테스트가 겹치기에 해당 테스트 함수를 제거 해주었습니다

GREENOVER reviewed (changes requested):
@bar-d 
안녕하세요 바드!
앞으로 2주간 찐 로직의 세계인 계산기에 오신걸 환영합니다🙌
그럼 잘 부탁드립니다🙃
우선 스텝 1 고생 많으셨어요ㅎㅎ
되게 빠르게 구현을 잘 해주셨고 아마 UML, TDD, 자료구조 등 낯설거나 처음 접해보는 개념들과 학습이 많았을텐데 전반적으로 개념 자체는 잘 이해하신것 같아 빠른 러닝커브를 지니신것 같다고 생각합니다! (부럽..)
그럼 우선 PR 코멘트에 대한 리뷰부터 하기전에 핫픽스로 처리해주셔야할 부분이 있어 공유드립니다!!
지금 현재 PR을 보내는 브랜치가 yagom-academy:main으로 되어있습니다ㅠㅠ
이 부분 바드에 해당 하는 브랜치로 수정해주시길 바랍니다😄
***
### 칭찬 드리고 싶은 부분
- 전체적으로 UML의 큼직 큼직한 그림들을 잘 설계해주셨습니다 💯
- 프로토콜화 시키고 채택해 사용하는점 아주 좋네요 💯
- 파일들을 관심사에 맞게 그룹화를 아주 잘 해주셨습니다 💯
- Queue 구현 방법에 대해 아주 많이 잘 찾아보셨고 성능 좋은 적절한 친구로 이유있게 잘 채택하셨네요 💯
- 간단한 개념이라도 딥하게 학습하시는 자세 아주 최고 💯💯💯
***
### UML에 대한 저의 의견
- 우선 전체적으로 UML을 보았을때 설계상 미스나 이해가 되지 않는 부분 없이 잘 나타내주셨어요!
하나 조금 더 확실히 표현하려면
```swift
<<struct>>
CalculatorItemQueue: Queue, CalculatorItem
```
처럼 채택되는 프로토콜을 명확하게 저런 시그니쳐로 표현하면 더 혼동되지 않는 UML이 될것 같았습니다!
### Queue에 대한 저의 의견
1. 우선 큐에서 꼭 필요한 기능들을 학습해보시고 기록해주신점 너무 좋아요..!!
다만 리뷰 코멘트에도 남겼지만 peek, isEmpty 같은것들이 대부분의 Queue 구현 시 기본적인 느낌이라곤 하지만 과연 우리 프로젝트에 접목시켰을때도 그럴지는 한번 생각해보면 좋을것 같습니다🙌
추후 제거하면 되는 별거 아닌거일 수 있지만 어떤 개념을 배우고 적용시킬때 정말 필요한 요소가 무엇인지 파악하고 이해하고 가져오는 학습이 병행되면 아주 나이스할것 같아요👍
2. Queue 구현에 대한 여러가지 방법을 아주아주아주 잘 학습하시고 또 기록해주신 부분들이 매우 맞습니다.
흔히 linkedList, DoubleStack으로 구현을 많이들 하는데 저또한 아직까지는 이중스택이 그나마 효율적이지 않나라고 생각합니다.
그런데 다른 구현 방법에는 단점들을 잘 파악해주셨는데 DoubleStack에서는 단점이 기재되어 있진 않아보여서 궁금했습니다!
그렇다면 바드가 생각하는 DoubleStack의 단점은 뭐가 있을까요?
단점없이 완전무결한 방법일까요..?🙋🏻
***
### 시간 복잡도에 대한 저의 의견
- 시간 복잡도까지 학습해보셨다니 아주 멋지십니다👍
전체적으로 정말 딥하고 여러 방면으로 학습을 하고 계시군요ㅎㅎ
현재 스텝에서 기재하신 시간 복잡도에 해당하는 array 메서드들 외에 다른 메서드 사용은 없을까요~?
(있을거니까 확인해보자!가 아니라 한번 더 체크하자는 의미였습니다😀)
그리고 번외로 append(contentsOf:)와 같은 친구의 시간 복잡도는 어떻게 될까요~?
***
### 궁금하신 부분에 대한 저의 의견
1. 학습해보셨다시피 Queue 타입이 딱 지정되어 사용할 수 있게 swift에서는 제공하지 않고 있고 여러 방법으로 구현하고 있습니다.
이에 DispatchQueue도 아마 까보시면 해당 용도로 Queue를 구현한것이고 지금 CalculatorItemQueue와 같이 큐의 역할을 한다~ 라고 이름을 붙여준 클래스일것입니다. 이를 생각해보면 해당 클래스의 존재여부와는 다를것 같고 큐의 구현이라고 보시면 될것 같아요!
2. 이해하신것과 같이 array도 List 타입으로 볼 수 있습니다.
즉 더 엄밀히 생각해보자면 List라는게 말씀하신것처럼 순서를 가진 추상된 타입의 개념이고 이를 표현하기 위해서 컬렉션 타입의 array를 사용한 DoubleStack이 될 수도 LinkedList가 될수도 있습니다.
3. TDD에 대해서 현재 구현해주신 스텝이 맞다고 생각합니다.
즉 UML로 모델타입에 대해 정의해주고 이를 테스트 코드를 구현하면서 지금처럼 값을 넣어 정상적으로 구현된 기능 정의인지 확인부터 해보는것이 TDD의 기초 개념이라 생각합니다!
이에 지금 전체적인 맥락에서 어긋나는 부분은 없다고 생각합니다😁
***
### 보완되면 좋을 부분
1. 전체적으로 리뷰 코멘트로도 많이 단 부분이 TDD 쪽인데 전체 흐름 및 개념에 대해서는 어긋나지 않았지만 그 검증을 위한 구현부가 조금 다듬어 지면 좋겠다는 생각이 있어 다시 한번 코멘트를 드립니다🙌
2. CalculatorItemQueue에 대해 조금 더 확장성 있는 방법을 생각해보면 좋을것 같습니다!
더불어 전체적으로 정의된 기능과 프로퍼티들의 실제 사용여부와 메서드 시그니쳐를 한번 더 고민해보면 어떨까해요🤔
3. 리드미를 매 스텝별 작성해보면 분명 분명 큰 도움이 되실거에요😀
4. 브랜치명이 조금 모호한것 같습니다.
혼자 작업할때는 문제가 되지 않지만 협업을 고려한다면 ic_6_barrd보다는 step1, step2와 같이 명확한 브랜치를 생성해 관리하는것은 어떨까요?
***
다시 한번 고생 많으셨습니다!
제 리뷰 코멘트들에 대해 다른 의견을 가지고 계시거나 더 나은 방향이 있다면 언제든 그 방향을 말씀해주시고 진행해주시면 됩니다😁
수정이 완료되면 리뷰 재요청을 해주세요ㅎㅎ
화이팅입니다🥳🥳🥳

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+//  Created by 김동용 on 2022/05/16.
+//
+
+//MARK: - CalculatorItemQueue
```
음 다른 struct들이 올 일은 아직 없어보입니다    
보통 다른 타입들이랑 구분해주기 위해 사용합니다!

bar-d committed:
feat: 폴더 및 파일 재정리

bar-d commented on Calculator/Calculator/Model/Protocol/Queue.swift:
```diff
@@ -0,0 +1,16 @@
+//  Created by 김동용 on 2022/05/16.
+//
+
+//MARK: - Queue
```
주석의 필요성이 없어보입니다🙃

bar-d commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        sut = nil
+    }
+    
+    func test_isEmpty_프로퍼티로_왼쪽과_오른쪽스택과배열이_모두비어있으면_true를_반환하는가() {
```
한글로 테스트명을 작성한다면 아무래도 어떤 테스트를 하고 있는지에 대한 명확성이 영어보다 좋은 단점이 있는것 같습니다  
단점은 한국에서만 쓸 수 있는 테스트코드가 아닐까요?🤔

bar-d committed:
style: 불필요한 주석 제거

bar-d committed:
feat: struct구조체 생성

bar-d commented on Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,7 @@
+//
+//  Calculator.swift
+//  Calculator
+//
+//  Created by 김동용 on 2022/05/16.
+//
+
```
나중에 유추하는데 더 도움이 될것 같습니다!

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+//
+
+//MARK: - CalculatorItemQueue
+struct CalculatorItemQueue<Double>: Queue, CalculateItem {  //double?
```
지운줄 알았는데 깜빡했네요🥲

bar-d committed:
refactor: CalculatorItemQueue 타입을 제너릭 타입으로 변경

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+//
+
+//MARK: - CalculatorItemQueue
+struct CalculatorItemQueue<Double>: Queue, CalculateItem {  //double?
```
처음 제네릭으로 타입을 지정해주었다가 숫자만 생각하고 테스트코드를 짜다보니 다시 Double 타입으로 바꿔주었던것 같습니다🥲  
다시 수정해주었습니다!

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+
+//MARK: - CalculatorItemQueue
+struct CalculatorItemQueue<Double>: Queue, CalculateItem {  //double?
+    var leftStack: Array<Double> = []
+    var rightStack: Array<Double> = []
```
그림으로 큐를 이해하다보니 왼쪽 스택과 오른쪽 스택이라는 네이밍을 하게 된것 같습니다.  
현재 네이밍에서 왼쪽 스택은 `enQueueStack`으로 바꿔주고 오른쪽 스택은`deQueueStack`으로 변경해주었습니다.

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+    
+    mutating func enQueue(_ double: Double) -> Bool {
```
`input`으로 네이밍을 변경해주었습니다

bar-d committed:
style: 네이밍 변경

bar-d committed:
refactor: xCode내 에러로 인해 info.plist삭제 후 폴더에 재추가

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+
+//MARK: - CalculatorItemQueue
+struct CalculatorItemQueue<Double>: Queue, CalculateItem {  //double?
+    var leftStack: Array<Double> = []
+    var rightStack: Array<Double> = []
```
접근제한자를 걸어 주었습니다!

bar-d commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+        var isEmpty: Bool { return (leftStack.isEmpty && rightStack.isEmpty) }
+        
+        //when
+        let result = true
```
처음 `XCTAssertEqual(result, isEmpty)`을 사용해 테스트를 하였는데,` XCTAssertTrue()`메소드를 사용해 보고 싶어서 사용했습니다  
이 과정에서 전에 사용했던 `result`를 지워주지 않은 것 같습니다🥲

bar-d commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+import XCTest
+@testable import Calculator
+
+class CalculatorTests: XCTestCase {
```
음... 현재 테스트는 정상적으로 실행되고 있습니다. 
`CalculatorTests` 모듈에서 `Calculatoer`의 타겟멤버쉽을 해제해주고 `CalculatorTests`만 체크해준 상태입니다.

GREENOVER replied:
흐음 신기하네요..!
Queue, CalculateItem의 타겟 멤버십에 Tests가 체크 되어 있지 않은데 이에 여파로 CalculatorItemQueue는 해당 두 프로토콜을 채택하고 있는데 이때문에 저는 바드의 프로젝트로 테스트 빌드 시 아래와 같이 뜨는데 다른 문제가 있는걸까요~?
![스크린샷 2022-05-17 오후 4 47 50](https://user-images.githubusercontent.com/72292617/168757983-3c4760e0-e0b5-43d4-8204-905c7fc105a0.png)

bar-d replied:
아 제가 커밋을 잘못 돌아갔습니다..!
그 이후에 커밋을 할 때 해당 모듈들의 타깃 멤버십을 다시 체크를 해주었었네요 
죄송합니다!

bar-d committed:
style: test코드 내 네이밍 변경

bar-d committed:
style: 불필요한 여백 제거

bar-d commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,144 @@
+@testable import Calculator
+
+class CalculatorTests: XCTestCase {
+    var sut: CalculatorItemQueue<Double>!
```
타입 자체를 옵셔널로 준다면 강제추출을 하지 않고 사용할 수 있을 것 같습니다

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+        return true
+    }
+    
+    mutating func deQueue() -> Double? {
```
제가 이해한게 맞다면 애초에 `enQueue`의 값을 넣어주는 스택에 값이 없다면 `deQueue`를 할수 없다는 말씀이실까요?  
그렇게 생각을 해보지도 못했네요😅  
더 좋은 방법인 것 같습니다!

GREENOVER replied:
네네 맞습니다!
그런 느낌이 되야하지 않을까해서요🙌

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+    
+    mutating func enQueue(_ double: Double) -> Bool {
```
처음 `UML`을 통해 코드를 구현하고 빨간색 에러가 나는게 싫어 임시로 넣어준 반환값이었는데, 빼주는 것을 까먹었습니다.
그런데 다시 생각해보니 현재 로직에서 반환값 자체가 필요가 없어보이네요

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+        return (leftStack.isEmpty && rightStack.isEmpty)
+    }
+    
+    var peek: Double? {
+        if !leftStack.isEmpty {
+            return leftStack.first
+        } else {
+            return rightStack.last
+        }
+    }
```
큐의 배열 내에서 첫번째 값을 알려주는 것을 확인하기 위해 peek이라는 프로퍼티를 통해 확인을 하게 해준다고 알고있습니다.
만약 `enQueue`메소드를 통해 넣어준 스택에 값이 있다면 `enQueue` 스택의 첫번째 값이 큐의 첫번째 값이 될 것이고, `deQueue`메소드를 통해 `deQueue`스택에 값이 저장이 되어있는 상태라면 `deQueue`스택의 마지막 값이 큐의 첫번째 값이 될 것입니다.   
만약 두 스택이 비어있다면 에러처리를 해줘야겠다는 생각이 드네요
해당 연산 프로퍼티는 계산기에서는 사용되지 않을것 같습니다.

GREENOVER replied:
네 그럴수도 있지만 deQueue에서 요소를 뽑아오고 그 해당 요소를 가지고 연산자/피연산자인지 구분을 할 수 있을것 같아서 실제적으로 사용을 안할수도 있겠다라는 생각이 들었어요!
물론 바드가 말씀하신것처럼 충분히 구현도 될거에요ㅎㅎ
이후 스텝에서 해보면서 어떤 방법이 더 클린할지 보면 좋겠습니다🙌

bar-d commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+    var leftStack: Array<Double> = []
+    var rightStack: Array<Double> = []
+    
+    var isEmpty: Bool {
+        return (leftStack.isEmpty && rightStack.isEmpty)
+    }
```
두 스택이 모두 비어있는 상태로 시작을 해야하기에 사용을 해야되는 것 같습니다

bar-d committed:
refactor: QueueEror열거형 생성 및 함수 리팩토링
- deQueue()함수 내 에러처리
- enQueue()함수 내 반환값 제거

bar-d committed:
refactor: peek이라는 프로퍼티가 반환하는 값을 검증하는 테스트 리팩토링

bar-d committed:
refactor: enQueue함수를 호출했을 때 왼쪽 스택에 값을 더하는지 검증하는 테스트코드 리팩토링

bar-d committed:
refactor: 테스트 코드 리팩토링 및 중복되는 테스트 코드 제거
- 컨벤션을 맞추기 위해 65번째 줄의 result, sut.enQueueStack.last 순서 변경
- 68번째 줄 의 테스트 코드 내 중복되는 코드 제거, 및 result 값 변경
- 중복되는 테스트 코드 제거

bar-d committed:
style: test함수 내 들여쓰기 통일

bar-d committed:
docs: README.md 파일 수정

bar-d committed:
refactor: 접근제한자 생성

bar-d committed:
refactor: Queue 프로토콜 및  CalculatorItemQueue 구조체 리팩토링
- 계산기 로직에 불필요한 프로퍼티 삭제
- 스택의 데이터를 없애주는 clearAllStacks 함수 생성

bar-d committed:
refactor: test코드 리팩토링

bar-d committed:
refactor: test코드 구현 완료 후 접근제한 설정

bar-d commented:
@GREENOVER   
일단 좋은 리뷰 감사드립니다 그린!  
너무 부족한 상태로 PR을 올린것 같아 부끄럽네요😅  
피드백 반영해서 다시 올려보았습니다!  
- 브랜치 같은 경우 그린의 피드백을 받고 step1이라는 브랜치를 생성하였는데, 현재 PR은 이미 ic_6_barrd라는 브랜치로 PR을 해서 체리픽을 사용해 피드백 반영한 커밋들을 옮겨주었습니다.
- 변경된 `UML`을 해당 PR에 수정해서 넣어주었고, README.md 파일에도 넣어주었습니다🙃

### `DoubleStack`
- 시간복잡도에 관련해 말씀해주신 `append(contentsOf:)`에도 관련이 있을 것 같은데요, `append(contentsOf:)`같은 경우 앞서 말씀드렸다 시피 평균 시간복잡도는 O(1)이지만, 메모리를 재할당 할 때 시간복잡도가 O(N)이 될 수 있기에 성능이 느려질 수도 있다는 단점이 있는 것 같습니다.
- 이 외에는 단점을 못찾았는데, 또 다른 단점이 있을까요?
- 시간 복잡도를 공부하다 공간 복잡도란 것도 있다는 것을 알게되었는데, 공간 복잡도에 관련된 것일까요? *제가 찾아본 바로는 `sorted()`와 `sort()` 관련된 이야기인 것 같아 아닌 것 같습니다.*


### `시간복잡도`
- 제가 시간 복잡도 관련 [참고한 블로그](https://demian-develop.tistory.com/30)에서는 `Array`, `Set`, `Dictionary`, `String`, `ClosedRange`, `고차함수`에 대한 시간복잡도에 대해 나와있었는데, 제가 PR에 표기한 시간복잡도들은 현재 프로젝트 내에서 사용한 메서드들에 대해서만 적어두었습니다.
- 시간복잡도 관련 궁금한게 하나가 있습니다.
- 시간복잡도를 고려해 `removeAll`이라는 메소드를 사용한다면 시간복잡도가 O(N)이고,   
  `removeLast`라는 메소드를 사용한다면 시간복잡도가 O(1)입니다.  
`removeAll`을 대신해 `removeLast`를 *n번* 반복한다면 결국 n으로 수렴할텐데,  
`removeLast`를 *n번* 반복하는 것과 `removeAll`을 한 번 사용하는 것의 차이가 있나요?


### `Queue의 프로퍼티 및 메서드`
- 큐의 기본저인 프로퍼티이긴 하지만 현재 계산기 프로젝트 내에서 생각을 해보기 위해 다시 STEP0의 계산기 구현설명을 다시 한번 보았습니다. 
- 해당 계산기의 로직을 살펴보면 `isEmpty`와 `peek`프로퍼티는 필요하지 않아보입니다.
- 또한 `AC`버튼을 누르게 되면 스택에 쌓인 모든 값을 지워줘야 하기에 `clearAllStacks`라는 메서드를 생성해 주었습니다.
- `clearAllStacks`메서드를 만들어 줄 때 `Array`의 메서드를 사용하지 않고 `enQueueStack = []`, `deQueueStack = []`을 넣어주었는데 시간복잡도 면에서 같을까요? 메서드를 사용하는게 더 좋을까요?  

### `TestCode`
- `TestCode`를 다 구현을 완료하고 접근제한을 걸어주어도 괜찮은가요?

GREENOVER commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,104 @@
+@testable import Calculator
+
+class CalculatorItemQueueTests: XCTestCase {
+    var sut: CalculatorItemQueue<Double>!
```
추후 기회가 되신다면 옵셔널로 만드셔보는것도 좋을것 같습니다🙋🏻

GREENOVER commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,104 @@
+        sut.deQueueStack = [1.0, 2.0, 3.0]
+        
+        //when
+        let result = try? sut.deQueue()
```
catch는 어디서 해주게 될까요~?🙋🏻

GREENOVER commented on README.md:
```diff
@@ -1,6 +1,180 @@
 ## iOS 커리어 스타터 캠프
 
-### 계산기 프로젝트 저장소
+### 계산기1
```
리드미 작성 너무 멋지십니다👍👍👍

GREENOVER commented on Calculator/Calculator/Model/Error/QueueError.swift:
```diff
@@ -0,0 +1,24 @@
+
+enum QueueError: LocalizedError {
+    case empty
+    case unknownError
```
보통은 unknown 정도로 표현합니다🙋🏻
이유는 QueueError안에서 다시 에러라는 중복된 네이밍을 가져갈 이유가 없다고 판단됩니다.
QueueError.unknwon일텐데 이미 unknown 자체가 에러를 내포한다고 생각해요😁

GREENOVER commented on Calculator/Calculator/Model/Error/QueueError.swift:
```diff
@@ -0,0 +1,24 @@
+}
+
+extension QueueError {
+    public var errorDescription: String? {
```
옵셔널 타입인 이유가 궁금합니다!

bar-d replied:
[문서](https://developer.apple.com/documentation/foundation/localizederror/2946895-errordescription)에서도 옵셔널값을 기본으로 하고 있고, 해당 에러에 대한 설명이 있을수도 있지만, 없을수도 있기 때문이 아닐까 생각합니다.

GREENOVER replied:
없을 수 있다는것이 궁금했어요..!
현재 QueueError에 대해 케이스를 명확히 지정해주셨고 연산 프로퍼티 안에서 return 값을 잘 지정해주셨기에 없을 수 있는값이 어떤게 들어올 수 있을지 궁금하네요🤔
문서에서 사용되는것은 단순 저 프로퍼티를 호출했을때는 옵셔널할 수는 있지만 지금 현재 구현에서는 옵셔널 할 수가 없지 않을까 생각이 듭니다🙋🏻

bar-d replied:
아 이해했습니다..!  
문서 내에서 옵셔널값을 프로퍼티로 주어서 일종의 컨벤션이라 생각한것 같습니다
제가 구현한 에러에서 description이 없으면 옵셔널로 처리하면 된다라는 말이군요

GREENOVER commented on Calculator/Calculator/Model/Error/QueueError.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+enum QueueError: LocalizedError {
```
에러 아주 잘 만들어 주셨네요👍

GREENOVER commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,31 @@
+    private(set) var enQueueStack: Array<T> = []
+    private(set) var deQueueStack: Array<T> = []
+    
+    public mutating func enQueue(_ input: T) {
+        enQueueStack.append(input)
+    }
+    
+    public mutating func deQueue() throws -> T? {
+        if enQueueStack.isEmpty {
+            throw QueueError.empty
+        }
+        else {
+            deQueueStack = enQueueStack.reversed()
+            enQueueStack.removeAll()
+        }
+        return deQueueStack.popLast()
+    }
+    
+    public mutating func clearAllStacks() {
```
public하게 해주신 이유가 궁금합니다🙋🏻

bar-d replied:
`public`은 선언한 모듈이 아닌 다른 모듈에서 사용이 가능한 접근레벨이고, `internal`은 기본 접근레벨로 해당 모듈 전체에서 사용이 가능합니다.  
두 접근레벨에 대해 잘못이해하고있었습니다.  
`internal`로 두어도 될것 같습니다 .

GREENOVER commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,31 @@
+        if enQueueStack.isEmpty {
+            throw QueueError.empty
+        }
+        else {
+            deQueueStack = enQueueStack.reversed()
+            enQueueStack.removeAll()
+        }
```
이미 if문에서 걸러지면 throw해서 빠져나가니까 else문을 제거해도 괜찮을것 같다고 생각해요🙋🏻

GREENOVER commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,31 @@
+        enQueueStack.append(input)
+    }
+    
+    public mutating func deQueue() throws -> T? {
```
옵셔널로 반환하는 이유가 있으실까요?🙋🏻

bar-d replied:
에러로 던져지면 리턴값이 없기에 옵셔널로 반환을 하였습니다

GREENOVER replied:
이 메서드를 호출하는쪽에서 try를 시도하고 catch에서 에러를 잡아줄텐데 만약 에러를 던져주게되면 리턴값은 자체적으로 반환되지 않지 않을까요??🙋🏻
즉 호출하는쪽에서 에러를 받아주는 분기를 타게되면 리턴값을 처리할 필요가 있을까합니다🤔

bar-d replied:
이해했습니다!  
리턴값을 반환해주지 않아도 에러로 분기를 타기에 리턴값 필요 없이 `deQueueStack.popLast()`를 호출해주면 되는거군요

GREENOVER reviewed (approved):
@bar-d 
리팩토링이 아주 잘되었네요!
고생 많으셨습니다..!
다시 한번 전체적으로 되돌아보시면서 더 보완된 부분들도 생기고 이번 스텝에서 많은걸 빠르게 잘 해보셨네요🙌
우선 PR 질문에 제 의견부터 드려보겠습니다ㅎㅎ
***
### DoubleStack
- 음 제가 알기로는 우선 append(contentsOf:)에 대한 평균 시간 복잡도는 O(n = element 수)라고 알고있어요..!
O(1)은 통상 1개의 Element가 들어올때일테고 만약 수천,수만개가 들어오게 된다면 O(수천,수만) 시간 복잡도를 가질것 같아요.
- 현재 계산기의 조건인 Queue의 구현에서만 한정되어 보자면 Stack으로의 구현에서 단점이 없다고 저도 생각해요!
다만 Stack의 공통된 점이겠지만 만약 필요에의해 어떤 값 다음 올 값을 찾을때 즉 해당 값을 찾아야된다면 처음부터 쭉 검사를 해야하니 시간복잡도가 조금 생길것 같아요. 딱 어떤 단점이 있어서 여쭤본건 아니였습니다🙋🏻
바드가 유추할때 어떤 단점들은 없었는지 궁금했었고 충분히 없을 수 있습니다ㅎㅎ
- 생각하셨다시피 공간 복잡도와는 관련없었습니다! 아마 제가 알기로는 DoubleStack이 공간복잡도도 좋았을거에요.
### 시간복잡도
- removeAll과 removeLast를 시간복잡도에서 보고 현재 계산기를 하는 딱 이 부분에 구현에서만 본다면 말씀하신것처럼 removeAll의 시간 복잡도는 O(N)이고 removeLast는 O(1)일거고 결국 반복을 돌려준다면 당연히 동일한 구현이 될거에요.
즉 바드가 말씀하신것처럼 현재 구현에서 두개를 어떻게 써도 성능과 역할의 차이는 없습니다.
그럼에도 removeAll을 현재에서 쓴다면 더 간결히 표현할 수 있다는 점이 있겠죠?
removeLast를 반복으로 돌릴 수도 있겠지만 removeAll이 있음에 그럴 필요가 없어지는거라 생각합니다🙋🏻
### Queue의 프로퍼티 및 메서드
- clearAllStack 구현에서 removeAll 혹은 빈 배열로 초기화에서 성능 차이는 제가 알기론 없습니다.
두 방법 모두 메모리 영향에 차이가 없습니다.
바드의 고민처럼 많은 사람들이 이 두개를 가지고 어떤게 더 낫냐 마냐 하던 레퍼가 있어 같이 공유드리겠습니다!
https://stackoverflow.com/questions/57788602/which-one-is-better-removeall-of-array-or-initialising-the-array-again
https://stackoverflow.com/questions/31183431/swift-delete-all-array-elements
아마 고민하시던거에 실마리를 얻을 수 있을거라 보입니다🙌
### Test
- 추후 다시 테스트를 하기 위해선 접근제어 즉 private(set) 설정을 풀어주는 등의 과정을 거쳐야하는걸까요?
그렇다면 한번 괜찮은 구조인지 생각해볼 필요가 있어보입니다..!
만약 정말 프라이빗하게 설정이되야하고 테스트를 꼭 해줘야한다면 외부에서 주입해서 사용하는 방식도 있을거에요.
그럼에도 불구하고 만약 잘 안된다면 구조가 맞는지를 다시 한번 보면 좋을거에요🙌
***
### 추후 스텝에서 보완되면 좋을 부분
- 현재는 테스트 케이스들이 모두 성공할때만을 고려해주셨는데 실제적으로 실패할때도 정상적으로 실패하는지에 대해서 실패 케이스를 같이 고려하고 테스트를 해준다면 더 확실한 검증이 될것 같습니다🙋🏻
***
현재 스텝에서 충분히 다음 스텝으로 넘어가기에 이상 없다고 판단됩니다ㅎㅎ
그렇기에 마이너한 리뷰 코멘트 및 의견을 여쭤보는것을 몇개 남겨두며 어프로브 > 머지 하겠습니다!
수고 많으셨어요ㅎㅎ
다음 스텝도 화이팅입니다🥳
p.s. 아..! 만약 혹시 리뷰 재요청을 해주실일이 있을때는 우측 상단에 저 리프레쉬 버튼을 눌러주시면 저한테 깃헙 알림이 옵니다ㅎㅎ
더 빠른 피드백을 받아보실 수 있을거에요!
물론 디코로 말씀해주셔도 되구요🙌
![스크린샷 2022-05-18 오전 9 53 21](https://user-images.githubusercontent.com/72292617/168935286-a0540eca-2ec4-462a-8bd1-e64fc0addbec.png)
